conans/client/plugin_manager.py

```diff
@@ -0,0 +1,125 @@
+
+attribute_checker_plugin = """
+def pre_export(output, conanfile, conanfile_path, reference, **kwargs):
+    # Check basic meta-data
+    for field in ["url", "license", "description"]:
+        field_value = getattr(conanfile, field, None)
+        if not field_value:
+            output.warn("Conanfile doesn't have '%s'. It is recommended to add it as attribute"
+                        % field)
+"""
+
+valid_plugin_methods = ["pre_export", "post_export",
+                        "pre_source", "post_source",
+                        "pre_build", "post_build",
+                        "pre_package", "post_package",
+                        "pre_upload", "post_upload",
+                        "pre_upload_recipe", "post_upload_recipe",
+                        "pre_upload_package", "post_upload_package",
+                        "pre_download", "post_download",
+                        "pre_download_recipe", "post_download_recipe",
+                        "pre_download_package", "post_download_package"
+                        ]
+
+
+class PluginManager(object):
+
+    def __init__(self, plugins_folder, plugin_names, output):
+        self._plugins_folder = plugins_folder
+        self._plugin_names = plugin_names
+        self.plugins = defaultdict(list)
+        self.output = output
+
+    def create_default_plugins(self):
+        attribute_checker_path = os.path.join(self._plugins_folder, "attribute_checker.py")
+        save(attribute_checker_path, attribute_checker_plugin)
+
+    def execute(self, method_name, **kwargs):
+        if not os.path.exists(os.path.join(self._plugins_folder, "attribute_checker.py")):
+            self.create_default_plugins()
+        if not self.plugins:
+            self.load_plugins()
+
+        assert method_name in valid_plugin_methods
+        for name, method in self.plugins[method_name]:
+            try:
+                output = ScopedOutput("[PLUGIN - %s] %s()" % (name, method_name), self.output)
+                method(output, **kwargs)
+            except Exception as e:
+                raise ConanException("[PLUGIN - %s] %s(): %s\n%s" % (name, method_name, str(e),
+                                                                     traceback.format_exc()))
+
+    def load_plugins(self):
+        for name in self._plugin_names:
+            self.load_plugin(name)
+
+    def load_plugin(self, plugin_name):
+        filename = "%s.py" % plugin_name
+        if filename and filename not in os.listdir(self._plugins_folder):
+            self.output.warn("Plugin '%s' not found in %s folder. Please remove plugin "
+                             "from conan.conf or include it inside the plugins folder."
+                             % (filename, self._plugins_folder))
+        else:
+            plugin_path = os.path.join(self._plugins_folder, filename)
+            for method in valid_plugin_methods:
+                try:
+                    plugin = PluginManager._load_module_from_file(plugin_path)
```

Comment:
Here we are loading the plugin one time for each method. Call the `plugin = PluginManager._load_module_from_file(plugin_path)` before the `for` loop.